README.md
[![Yii2](https://img.shields.io/badge/required-Yii2_v2.0.40-blue.svg)](https://packagist.org/packages/yiisoft/yii2)
[![Downloads](https://img.shields.io/packagist/dt/wdmg/yii2-forms.svg)](https://packagist.org/packages/wdmg/yii2-forms)
[![Packagist Version](https://img.shields.io/packagist/v/wdmg/yii2-forms.svg)](https://packagist.org/packages/wdmg/yii2-forms)
![Progress](https://img.shields.io/badge/progress-ready_to_use-green.svg)
[![GitHub license](https://img.shields.io/github/license/wdmg/yii2-forms.svg)](https://github.com/wdmg/yii2-forms/blob/master/LICENSE)

<img src="./docs/images/yii2-forms.png" width="100%" alt="Yii2 Composite Forms" />

# Yii2 Forms
Module for creating and administering composite forms for Yii2. It also records and manages the results of filling out forms on the site. It has its own component for building and displaying forms in the frontend. Supports multilingual fields and descriptions.

This module is an integral part of the [Butterfly.СMS](https://butterflycms.com/) content management system, but can also be used as an standalone extension.

Copyrights (c) 2019-2021 [W.D.M.Group, Ukraine](https://wdmg.com.ua/)

# Requirements 
* PHP 5.6 or higher
* Yii2 v.2.0.40 and newest
* [Yii2 Base](https://github.com/wdmg/yii2-base) module (required)

# Installation
To install the module, run the following command in the console:

`$ composer require "wdmg/yii2-forms"`

After configure db connection, run the following command in the console:

`$ php yii forms/init`

And select the operation you want to perform:
  1) Apply all module migrations
  2) Revert all module migrations
  3) Add demo data

# Migrations
In any case, you can execute the migration and create the initial data, run the following command in the console:

`$ php yii migrate --migrationPath=@vendor/wdmg/yii2-forms/migrations`

# Configure

To add a module to the project, add the following data in your configuration file:

    'modules' => [
        ...
        'forms' => [
            'class' => 'wdmg\forms\Module',
            'routePrefix' => 'admin'
        ],
        ...
    ],

# Usage examples
To build the ActiveForm with fields you may use the component method Yii::$app->forms->build() with `id` or `alias` of form item.

**View in frontend**

    <?php
        use yii\widgets\ActiveForm;
        ...
    ?>
        
    <h3>Feedback</h3>
    <?php $form = ActiveForm::begin(); ?>
    <?= Yii::$app->forms->build($form, 'feedback-form'); ?>
    <?= Html::submitButton('Submit') ?>
    <?php ActiveForm::end(); ?>
    ...
    
    <h3>Callback</h3>
    <?php $form = ActiveForm::begin(); ?>
    <?= Yii::$app->forms->build($form, 'сallback-form'); ?>
    <?= Html::submitButton('Submit') ?>
    <?php ActiveForm::end(); ?>

**Action in Controller**

    <?php
        if (!is_null($result = Yii::$app->forms->submit('feedback-form', $data))) {
            if ($result === true) {
                Yii::$app->session->setFlash('success', 'Your Feedback form successfully submitted!');
            } else {
                Yii::$app->session->setFlash('danger', 'An error occurred while sending the Feedback form.');
            }
        }
    ?>

# Routing
Use the `Module::dashboardNavItems()` method of the module to generate a navigation items list for NavBar, like this:

    <?php
        echo Nav::widget([
        'options' => ['class' => 'navbar-nav navbar-right'],
            'label' => 'Modules',
            'items' => [
                Yii::$app->getModule('forms')->dashboardNavItems(),
                ...
            ]
        ]);
    ?>


# Status and version [ready to use]
* v.1.2.0 - Update copyrights, fix menu dashboard
* v.1.1.2 - RBAC implementation
* v.1.1.1 - Update README.md and setAttributeLabels() fixed


CHANGELOG.md
Changelog
=========

## 1.2.0 (2023-07-06)
 * Update copyrights, fix menu dashboard
 * Lang attribute for HTML tags and inputs
 
## 1.1.2 (2020-12-28)
 * RBAC implementation

## 1.1.1 (2020-06-14)
 * Update README.md
 * setAttributeLabels() fixed

## 1.1.0 (2020-05-03)
 * Complete views of submitted forms
 * Added submit() method for Form component
 * SluggableBehavior for Fields name
 * Added flags and language labels
 * Added demo data
 * Multi-language support, migrations
 
## 1.0.13 (2020-04-14)
 * Update dependencies
 
## 1.0.12 (2020-04-01)
 * Log activity and translations
 * Added forms component
 
## 1.0.11 (2020-03-26)
 * Added pagination
 * Up to date dependencies
 
## 1.0.10 (2020-03-22)
 * Rebuild migrations, views
 * Added translations
 
## 1.0.9 (2019-12-10)
 * Fixed deprecated class declaration

## 1.0.8 (2019-07-16)
 * Added extra options to composer.json and navbar menu icon

## 1.0.7 (2019-06-09)
 * Added choice param for non interactive mode
 
## 1.0.6 (2019-06-07)
 * Module refactoring
 
## 1.0.5 (2019-06-05)
 * Module transferred to base module interface. Update Yii2 version.

## 1.0.4 (2019-06-01)
 * Bugfix, routing and refactoring

## 1.0.3 (2019-05-31)
 * Update Yii2 version
 
## 1.0.2 (2019-05-28)
 * Added views, module name and description, translations. Routing fix.
 
## 1.0.1 (2019-03-16)
 * Added dashboard navigation items for NavBar

## 1.0.0 (2019-03-08)
 * Initial

models/Fields.php
<?php

namespace wdmg\forms\models;

use Yii;
use wdmg\base\models\ActiveRecordML;
use wdmg\forms\models\Forms;
use yii\behaviors\BlameableBehavior;
use yii\behaviors\TimestampBehavior;
use yii\db\ActiveRecord;
use yii\db\Expression;
use yii\helpers\ArrayHelper;
use wdmg\base\behaviors\SluggableBehavior;

/**
 * This is the model class for table "{{%forms_fields}}".
 *
 * @property int $id
 * @property int $form_id
 * @property string $name
 * @property string $label
 * @property string $description
 * @property int $type
 * @property int $sort_order
 * @property string $params
 * @property int $is_required
 * @property int $status
 *
 * @property FormsContent[] $formsContents
 * @property FormsList $form
 */
class Fields extends ActiveRecordML
{

    const STATUS_DRAFT = 0; // Form field has draft
    const STATUS_PUBLISHED = 1; // Form field has been published

    private $fieldsTypes = [
        1 => 'text',
        2 => 'textarea',
        3 => 'checkbox',
        4 => 'file',
        5 => 'hidden',
        6 => 'password',
        7 => 'radio',
        8 => 'color',
        9 => 'date',
        10 => 'datetime',
        11 => 'datetime-local',
        12 => 'email',
        13 => 'number',
        14 => 'range',
        15 => 'search',
        16 => 'tel',
        17 => 'time',
        18 => 'url',
        19 => 'month',
        20 => 'week'
    ];

    public $attribute;

    /**
     * {@inheritdoc}
     */
    public static function tableName()
    {
        return '{{%forms_fields}}';
    }

    /**
     * {@inheritdoc}
     */
    public function behaviors()
    {
        $behaviors = parent::behaviors();
        $behaviors['sluggable'] = [
            'class' => SluggableBehavior::class,
            'attribute' => 'label',
            'slugAttribute' => 'name',
            'replacement' => '_'
        ];
        return $behaviors;
    }

    /**
     * {@inheritdoc}
     */
    public function rules()
    {
        $rules = [
            [['form_id', 'label', 'type', 'status'], 'required'],
            [['form_id', 'type', 'sort_order', 'status'], 'integer'],
            [['params'], 'string'],
            [['label', 'name'], 'string', 'max' => 64],
            ['name', 'match', 'pattern' => '/^[A-za-z]/', 'message' => Yii::t('app/modules/forms','The attribute must begin with a letter.')],
            ['name', 'match', 'pattern' => '/^[A-Za-z0-9\_]+$/', 'message' => Yii::t('app/modules/forms','It allowed only Latin alphabet, numbers and «_» character.')],
            [['placeholder'], 'string', 'max' => 124],
            [['description'], 'string', 'max' => 255],
            [['form_id'], 'exist', 'skipOnError' => false, 'targetClass' => Forms::class, 'targetAttribute' => ['form_id' => 'id']],
            [['created_at', 'updated_at'], 'safe'],
        ];

        if (class_exists('\wdmg\users\models\Users')) {
            $rules[] = [['created_by', 'updated_by'], 'safe'];
        }

        return $rules;
    }

    /**
     * {@inheritdoc}
     */
    public function attributeLabels()
    {
        return [
            'id' => Yii::t('app/modules/forms', 'ID'),
            'form_id' => Yii::t('app/modules/forms', 'Form ID'),
            'label' => Yii::t('app/modules/forms', 'Field label'),
            'name' => Yii::t('app/modules/forms', 'Input name'),
            'placeholder' => Yii::t('app/modules/forms', 'Placeholder'),
            'description' => Yii::t('app/modules/forms', 'Description'),
            'type' => Yii::t('app/modules/forms', 'Type'),
            'sort_order' => Yii::t('app/modules/forms', 'Sort order'),
            'params' => Yii::t('app/modules/forms', 'Params'),
            'is_required' => Yii::t('app/modules/forms', 'Is required?'),
            'status' => Yii::t('app/modules/forms', 'Status'),
            'created_at' => Yii::t('app/modules/forms', 'Created at'),
            'created_by' => Yii::t('app/modules/forms', 'Created by'),
            'updated_at' => Yii::t('app/modules/forms', 'Updated at'),
            'updated_by' => Yii::t('app/modules/forms', 'Updated by'),
        ];
    }

    /**
     * {@inheritdoc}
     */
    public function afterFind()
    {
        parent::afterFind();
        $this->attribute = str_replace('-', '_', $this->name);
    }

    /**
     * @return \yii\db\ActiveQuery
     */
    public function getFormsContents()
    {
        return $this->hasMany(Content::class, ['input_id' => 'id']);
    }

    /**
     * @return \yii\db\ActiveQuery
     */
    public function getForm()
    {
        return $this->hasOne(Forms::class, ['id' => 'form_id']);
    }

    /**
     * @return array
     */
    public function getFieldsTypes()
    {
        return $this->fieldsTypes;
    }

    /**
     * @return array
     */
    public function getFieldType($type = null)
    {
        if (is_null($type))
            return null;

        $types = $this->getFieldsTypes();
        if (isset($types[$type]))
            return $types[$type];
        else
            return null;
    }

    /**
     * @return array
     */
    public function getFieldsTypesList($allTypes = false)
    {
        $list = [];
        if ($allTypes)
            $list['*'] = Yii::t('app/modules/forms', 'All types');

        $list = ArrayHelper::merge($list, $this->getFieldsTypes());

        return $list;
    }

    /**
     *
     */
    public function getValidator()
    {
        switch ($this->type) {
            case 1: // 'text'
                return 'string';
            case 2: // 'textarea'
                return 'string';
            case 3: // 'checkbox'
                return 'string';
            case 4: // 'file'
                return 'string';
            case 5: // 'hidden'
                return 'string';
            case 6: // 'password'
                return 'string';
            case 7: // 'radio'
                return 'string';
            case 8: // 'color'
                return 'string';
            case 9: // 'date'
                return 'string';
            case 10: // 'datetime'
                return 'string';
            case 11: // 'datetime-local'
                return 'string';
            case 12: // 'email'
                return 'string';
            case 13: // 'number'
                return 'string';
            case 14: // 'range'
                return 'string';
            case 15: // 'search'
                return 'string';
            case 16: // 'tel'
                return 'string';
            case 17: // 'time'
                return 'string';
            case 18: // 'url'
                return 'string';
            case 19: // 'month'
                return 'string';
            case 20: // 'week'
                return 'string';
            default:
                return 'string';
        }
    }

    /**
     * @return object of \yii\db\ActiveQuery
     */
    public function getAllForms($cond = null, $select = ['id', 'name'], $asArray = false)
    {
        if ($cond) {
            if ($asArray)
                return Forms::find()->select($select)->where($cond)->asArray()->indexBy('id')->all();
            else
                return Forms::find()->select($select)->where($cond)->all();

        } else {
            if ($asArray)
                return Forms::find()->select($select)->asArray()->indexBy('id')->all();
            else
                return Forms::find()->select($select)->all();
        }
    }

    /**
     * @return array
     */
    public function getAllFormsList($allTags = false)
    {
        $list = [];
        if ($allTags)
            $list['*'] = Yii::t('app/modules/forms', 'All forms');

        if ($tags = $this->getAllForms(['source_id' => null], ['id', 'name'], true)) {
            $list = ArrayHelper::merge($list, ArrayHelper::map($tags, 'id', 'name'));
        }

        return $list;
    }

    /**
     * @return array
     */
    public function getIsRequiredList($allVariants = false)
    {
        $list = [];
        if ($allVariants) {
            $list = [
                '*' => Yii::t('app/modules/forms', 'All variants')
            ];
        }

        return ArrayHelper::merge($list, [
            0 => Yii::t('app/modules/forms', 'Required'),
            1 => Yii::t('app/modules/forms', 'Not required'),
        ]);
    }

    /**
     * @return array
     */
    public function getStatusesList($allStatuses = false)
    {
        if ($allStatuses)
            return [
                '*' => Yii::t('app/modules/forms', 'All statuses'),
                self::STATUS_DRAFT => Yii::t('app/modules/forms', 'Draft'),
                self::STATUS_PUBLISHED => Yii::t('app/modules/forms', 'Published'),
            ];
        else
            return [
                self::STATUS_DRAFT => Yii::t('app/modules/forms', 'Draft'),
                self::STATUS_PUBLISHED => Yii::t('app/modules/forms', 'Published'),
            ];
    }
}
